metronome/hotstuff/service/test/src/io/iohk/metronome/hotstuff/service/execution/BlockExecutorProps.scala

```diff
@@ -0,0 +1,287 @@
+              viewNumber = viewNumber,
+              blockHash = descendant.id,
+              signature = GroupSignature(())
+            )
+            effect = Effect.ExecuteBlocks[TestAgreement](
+              lastExecutedBlockHash = ancestor.id,
+              quorumCertificate = commitQC
+            )
+          } yield (tree ++ descendantTree, effects :+ effect)
+
+          extension.flatMap { case (tree, effects) =>
+            loop(i - 1, tree, effects)
+          }
+        }
+      }
+
+      for {
+        prefixTree <- genNonEmptyBlockTree
+        i          <- Gen.choose(minBatches, maxBatches)
+        fixture    <- loop(i, prefixTree, Vector.empty)
+      } yield fixture
+    }
+  }
+
+  def run(test: Task[Prop]): Prop = {
+    import Scheduler.Implicits.global
+    test.runSyncUnsafe(timeout = 5.seconds)
+  }
+
+  property("executeBlocks - from root") = forAll { (fixture: TestFixture) =>
+    run {
+      fixture.resources.use { case (blockSychronizer, _) =>
+        for {
+          _ <- fixture.batches.traverse(blockSychronizer.enqueue)
+
+          executedBlockHashes <- fixture.awaitBlockExecution(
+            fixture.lastBatchCommitedBlockHash
+          )
+
+          // The genesis was the only block we marked as executed.
+          pathFromRoot <- fixture.storeRunner.runReadOnly {
+            TestBlockStorage.getPathFromRoot(fixture.lastBatchCommitedBlockHash)
+          }
+
+        } yield {
+          "executes from the root" |: executedBlockHashes == pathFromRoot.tail
+        }
+      }
+    }
+  }
+
+  property("executeBlocks - from last") = forAll { (fixture: TestFixture) =>
+    run {
+      fixture.resources.use { case (blockSychronizer, viewStateStorage) =>
+        val lastBatch             = fixture.batches.last
+        val lastExecutedBlockHash = lastBatch.lastExecutedBlockHash
+        for {
+          _ <- fixture.storeRunner.runReadWrite {
+            viewStateStorage.setLastExecutedBlockHash(lastExecutedBlockHash)
+          }
+          _ <- blockSychronizer.enqueue(lastBatch)
+
+          executedBlockHashes <- fixture.awaitBlockExecution(
+            fixture.lastBatchCommitedBlockHash
+          )
+
+          pathFromLast <- fixture.storeRunner.runReadOnly {
+            TestBlockStorage.getPathFromAncestor(
+              lastExecutedBlockHash,
+              fixture.lastBatchCommitedBlockHash
+            )
+          }
+
+        } yield {
+          "executes from the last" |: executedBlockHashes == pathFromLast.tail
+        }
+      }
+    }
+  }
+
+  property("executeBlocks - from pruned") = forAll { (fixture: TestFixture) =>
+    run {
+      fixture.resources.use { case (blockSychronizer, _) =>
+        val lastBatch             = fixture.batches.last
+        val lastExecutedBlockHash = lastBatch.lastExecutedBlockHash
+        for {
+          _ <- fixture.storeRunner.runReadWrite {
+            TestBlockStorage.pruneNonDescendants(lastExecutedBlockHash)
```

Comment:
I don't quite understand this test. Do we even populate the storage with such non-descendants?

Reply:
Sorry, there aren't many comments in the setup.

Yeah, we generate a [prefixTree](https://github.com/input-output-hk/metronome/pull/39/files/7fd73c729ad087906359aa78ad637b9ce3b41108#diff-43a00c2f1553bcab8cd7e482b1c3d5861d3cc10c91528c904910a85a46e909f4R176) first and then between 1 and 5 extension trees; this test picks `lastBatch.lastExecutedBlockHash` as the new root, so after pruning we shall only have the last extension tree. The non-descendants are the ancestors, uncles and cousins in the prefix tree and any previous extensions.

The test checks that even if we prune the block that were the last executed ones according to our view state, the `BlockExecutor` can recover by going from whatever the last committed block before the current batch was.

This is implemented [here](https://github.com/input-output-hk/metronome/pull/39/files/7fd73c729ad087906359aa78ad637b9ce3b41108#diff-6c7370a74c3203456a648dd0b2eb746f549b452dd00006152771767883fba0dbR111-R114) but I can't remember why I thought a situation like this would arise. I only implemented pruning for the robot and it prunes blocks [relative to the last executed hash](https://github.com/input-output-hk/metronome/pull/42/files#diff-10b56ac2536b604ca590b5fe6859da38983d0023e5199eb3d5659d1d9cee12ddR419-R421), so this shouldn't happen. Maybe it was before I added the reading of the last hash from the view state before each batch :thinking:

Reply:
Yeah, that's how it [started](https://github.com/input-output-hk/metronome/pull/39/commits/50493682ce2decf005352e772bdb178c8529aabb#diff-6c7370a74c3203456a648dd0b2eb746f549b452dd00006152771767883fba0dbR43-R56), and I think I was just being defensive, thinking what would happen if someone would prune away the blocks we thought we last executed, surely they would keep at least the last committed one.

Reply:
![plans within plans](https://memegenerator.net/img/instances/74103097.jpg)

Reply:
I'm still not quite clear. If I understand correctly the initial store is a chain of blocks, so what is there to be pruned?

Reply:
The initial store has all the block tree that was generated, and the batches are what will tell what the committed chain from that is. 

Say it has a tree like this were `0*` is the `prefixTree`, `1*` is the first and `2*` is the second extension:
```
                            15
                          /
01 - 02 - 04 - 11 - 12 - 14 - 16 
   \                 \ 
    03                13 - 17 - 21 - 22 - 23
                                      \
                                       24 - 25
```
Then the fixture would have 2 batches: `(04, 17)` and `(17, 25)`. In each pair the 1st member is what has been committed in the previous batch, the 2nd is the new committed block. The pruning in this case would get rid of anything before `17`. 

> the call to pruneNonDescendants is kind of arbitrary. What if it's called with a different blockHash than what we expect?

You are right, the pruning here is specific to one of the conditions that the `BlockExecutor` knows how to handle. If we pruned from `13`, it wouldn't execute anything. The batches mean that we only have commits for `04`, `17` and `25`. The recovery mechanism in the executor works by assuming that even if something went wrong during the execution, the federation will keep working and give us a new commit at some point, at which point we can again try and see if we can find that block. 

In our example, the `lastExecutedBlockHash` would be `01` but for some reason someone pruned to `13`, so the batch of `(04, 17)` would fail because neither `01` nor `04` exist. But then would come batch `(17, 25)` and we would see that `17` indeed exists in storage. We ask the `ApplicationService` to execute `21`, `22`, `24` and `25`, because it always assumes that the first item of the batch has been done before. If `21` fails, then we are stuck, but if we indeed synced to `17` somehow, then it can resume from there  and update `lastExecutedBlockHash`.

The failure scenario is tested further down. This test only checks that if the `lastExecutedHash` in `ViewStateStorage` doesn't exist but the one in the batch does, we use that one to attempt to execute from. 

> maybe we should add a test for the executionSkipped case?

Yeah, good idea, will add one :+1:

Reply:
Added the test for `executionSkipped` in https://github.com/input-output-hk/metronome/pull/39/commits/40d1a00a78120627012c88cee222526062bb5a8b

Thank you for raising it, it caught a bug as well, the first block that failed the compare-and-set wasn't traced.